boards/shared/Src/App/App_SharedStateMachine.c

```diff
@@ -10,6 +11,38 @@ struct StateMachine
     struct World *      world;
 };
 
+/**
+ * Run the given tick function over the given state machine if the tick function
+ * is not null
+ *
+ * @param state_machine The state machine to run the tick function over
+ * @param tick_function The tick function to run over the state machine
+ */
+void App_SharedStateMachine_RunStateTickFunctionIfNotNull(
```

Comment:
I was a little bit worried about whether changing the state machine to be multithreaded cause any race condition. One potential race condition I could think of:


Assuming the following:
1. We have `OnTick1Khz` and `OnTick1Hz`
2. `Task1Khz` is higher priority than `Task1Hz`
3. We have two states - `INIT` and `FAULT`, and we default to `INIT`

```
OnTick1Khz() {
  SetNextState(INIT);
}

OnTick1Hz() {
  SetNextState(FAULT);
  // An task switch to Task1kHz  occurs before we return from this function
}
```

It seems to me that if `OnTick1Hz` requests a transition to `FAULT` state and then gets preempted by `OnTick1kHz` (which overrides NexState to be `INIT`), then we never transition to `FAULT` state. Obviously this example is contrived but I think you get the scenario I am trying to describe.

Reply:
Ok, added a semaphore lock here that should prevent this.

Reply:
That's quite a big section to guard but I've got no better idea at the moment so let's run with it. I think we should leave the semaphore inside `struct StateMachine` though?

Also you need to initialize it: https://www.freertos.org/xSemaphoreCreateMutexStatic.html

Reply:
Good catch, added.

Reply:
Looks like the semaphore is initialized properly now, but it's still not part of `struct StateMachine`. Was that a deliberate choice? It seems to make more sense to have a per-state machine mutex. Suppose we ever have multiple state machines, you would want an unique mutex for each of them?

Reply:
Whoops, I did actually get part way through doing that design, but bamboozled myself into thinking each mutex needed a static storage duration..... anyhow you're totally right. Moved the mutex and it's storage to the `StateMachine` struct and changed everything else to work with that. 

:+1: